### PR TITLE
Upgrade and generalise basis state preparation

### PR DESCRIPTION
**Pull request by KetpuntoG**
This PR complete part of this story: [[sc-68521](https://app.shortcut.com/xanaduai/story/68521)]

Goal: `BasisEmbedding` is an alias of `BasisState`. This way, we don't have duplicate code that does the same thing.
In unifying this, I have had to modify some tests due to:
- `BasisEmbedding` and `BasisState` throw errors such as "incorrect length" with different messages. Now it will always be the same. (test modified for this reason: `test_default_qubit_legacy.py`, `test_default_qubit_tf.py`
`test_default_qubit_torch.py`, `test_state_prep.py`, `test_all_singles_doubles.py` and test_uccsd`)

- In `BasisEmbedding`, errors were thrown in `__init__` while in BasisState in `state_vector`. Now they are unified in `__init__`. For this reason, there were tests where the operator was not initialized correctly but no error was thrown since `state_vector` was not being called but now they are detected. To correct this, I have modified the tests: `test_qscript.py`, `test_state_prep.py`,

- Now `BasisState` does not decompose `BasisStatePreparation` since we are going to deprecate it. This causes the number of gates after expanding to be affected. In this case I had to modify some test in `test_tape.py`.

This PR also solves:

- [issue 6008](https://github.com/PennyLaneAI/pennylane/issues/6008)
- [issue 6007](https://github.com/PennyLaneAI/pennylane/issues/6007)
- [issue 6006](https://github.com/PennyLaneAI/pennylane/issues/6006)

### CODE REVIEW

**Comment by KetpuntoG** [comment]
[sc-68521]

File: pennylane/templates/embeddings/basis.py [hunk 1]
```diff
@@ -15,146 +15,10 @@
...
-
-        ops_list = []
-        for wire, state in zip(wires, basis_state):
-            ops_list.append(qml.PhaseShift(state * np.pi / 2, wire))
-            ops_list.append(qml.RX(state * np.pi, wire))
-            ops_list.append(qml.PhaseShift(state * np.pi / 2, wire))
-
-        return ops_list
+# pylint: disable=missing-class-docstring
+class BasisEmbedding(BasisState):
```
Comment (trbromley) [inline pennylane/templates/embeddings/basis.py:23]: Are we able to keep the docstrings? It might be nice to maintain the more QML-focused positioning.

Comment (KetpuntoG) [reply pennylane/templates/embeddings/basis.py:23]: Hmm I think it should work 👌

File: doc/development/deprecations.rst [hunk 2]
```diff
@@ -63,6 +63,12 @@ Pending deprecations
   - Deprecated in v0.37
   - Will be removed in v0.39
 
+* The ``BasisStatePreparation`` template is deprecated.
+  Instead, ``BasisState`` can be called on the constructed operator.
+
+  - Deprecated in v0.38
+  - Will be removed in v0.39
```
Comment (KetpuntoG) [inline doc/development/deprecations.rst:70]: Double check

Comment (trbromley) [reply doc/development/deprecations.rst:86]: Works for me! Any concerns @isaacdevlugt?

Comment (isaacdevlugt) [reply doc/development/deprecations.rst:86]: Thanks @KetpuntoG! I think it's fine to deprecate this for v0.38, remove for v0.39. I created a story in the deps / removals for v0.38 that relates to the story Guillermo is working on :)

File: doc/development/deprecations.rst [hunk 3]
```diff
@@ -79,6 +79,12 @@ Pending deprecations
   - Deprecated in v0.37
   - Will be removed in v0.39
 
+* The ``BasisStatePreparation`` template is deprecated.
+  Instead, ``BasisState`` can be called on the constructed operator.
```
Comment (isaacdevlugt) [inline doc/development/deprecations.rst:83]:
```suggestion
* The ``BasisStatePreparation`` template is deprecated.
  Instead, use ``BasisState``.
```

File: patata.py [hunk 4]
```diff
@@ -0,0 +1,3 @@
+import pennylane as qml
```
[the remark this answers is not in the document]
Comment (KetpuntoG) [reply patata.py:1]: Oh no 🙈

File: pennylane/devices/tests/test_templates.py [hunk 5]
```diff
@@ -225,6 +225,15 @@ def circuit():
...
+    @pytest.fixture(scope="function", autouse=True)
+    def capture_warnings(self, recwarn):
+        """Capture warnings."""
+        yield
+        if len(recwarn) > 0:
+            for w in recwarn:
+                assert isinstance(w.message, qml.PennyLaneDeprecationWarning)
+                assert "BasisStatePreparation is deprecated" in str(w.message)
+
     def test_BasisStatePreparation(self, device, tol):
```
Comment (albi3ro) [inline pennylane/devices/tests/test_templates.py:237]: We can just update this test to use `qml.BasisState` instead of `qml.BasisStatePreparation`.

File: pennylane/devices/tests/test_templates.py [hunk 6]
```diff
@@ -225,6 +225,15 @@ def circuit():
...
         )
 
+    @pytest.fixture(scope="function", autouse=True)
+    def capture_warnings(self, recwarn):
+        """Capture warnings."""
+        yield
+        if len(recwarn) > 0:
+            for w in recwarn:
+                assert isinstance(w.message, qml.PennyLaneDeprecationWarning)
+                assert "BasisStatePreparation is deprecated" in str(w.message)
```
Comment (astralcai) [inline pennylane/devices/tests/test_templates.py:235]: This fixture captures all warnings in all tests and assert that they are raised from `BasisStatePreparation`, including warnings raised from deprecated devices. Maybe instead of a fixture, use `with pytest.raises` in the individual tests that uses the deprecated features?

Comment (KetpuntoG) [reply pennylane/devices/tests/test_templates.py:235]: The error seems to pop up in all tests regardless of whether they use BasisStatePreparation or not. So I don't think the problem is in the decorator 🤔 Maybe this test is called from somewhere that instantiates the template?

Comment (KetpuntoG) [reply pennylane/devices/tests/test_templates.py:235]: or maybe you are right 😆

File: doc/releases/changelog-dev.md [hunk 7]
```diff
@@ -310,7 +314,16 @@
...
+  [(#5969)](https://github.com/PennyLaneAI/pennylane/pull/5969)
+
+* `qml.BasisState` now works with jax.jit.
+  [(#6021)](https://github.com/PennyLaneAI/pennylane/pull/6021)
+
+* `qml.BasisEmbedding` now gives the correct decomposition.
+  [(#6021)](https://github.com/PennyLaneAI/pennylane/pull/6021)
+
+* `qml.BasisEmbedding` now works with lightning.qubit and jax.jit.
+  [(#6021)](https://github.com/PennyLaneAI/pennylane/pull/6021)
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:326]: Not necessarily a blocker, but this is a potential sign that the PR is doing too much.

Comment (KetpuntoG) [reply doc/releases/changelog-dev.md:326]: I agree, however in this case it came naturally when reformatting the classes

File: pennylane/devices/tests/test_templates.py (hunk 5 above)
Comment (KetpuntoG) [reply pennylane/devices/tests/test_templates.py:237]: Solved with `autouse = False` 😄

**Comment by KetpuntoG** [comment]
[sc-68521]

File: doc/releases/changelog-dev.md (hunk 7 above)
Comment (soranjh) [reply doc/releases/changelog-dev.md:329]: Agree with Tom, this PR is too long.

File: tests/ops/qubit/test_state_prep.py [hunk 8]
```diff
@@ -351,18 +351,9 @@ def test_BasisState_state_vector_bad_wire_order(self):
         with pytest.raises(WireError, match="wire_order must contain all BasisState wires"):
             basis_op.state_vector(wire_order=[1, 2])
 
-    def test_BasisState_explicitly_checks_0_1(self):
```
Comment (KetpuntoG) [inline tests/ops/qubit/test_state_prep.py:354]: With the new functionality it can also receive integers

File: pennylane/ops/qubit/state_preparation.py [hunk 9]
```diff
@@ -20,24 +20,24 @@
...
+from pennylane.templates.state_preparations import MottonenStatePreparation
 from pennylane.typing import TensorLike
 from pennylane.wires import WireError, Wires, WiresLike
 
 state_prep_ops = {"BasisState", "StatePrep", "QubitDensityMatrix"}
 
 
 class BasisState(StatePrepBase):
-    r"""BasisState(n, wires)
+    r"""BasisState(features, wires)
```
Comment (soranjh) [inline pennylane/ops/qubit/state_preparation.py:34]: Probably better to keep `n` or at least change it to `state` for consistency with `StatePrep`.

File: pennylane/ops/qubit/state_preparation.py [hunk 10]
```diff
@@ -96,33 +131,49 @@ def compute_decomposition(n: TensorLike, wires: WiresLike) -> list[Operator]:
...
+        return op_list
 
     def state_vector(self, wire_order: Optional[WiresLike] = None) -> TensorLike:
         """Returns a statevector of shape ``(2,) * num_wires``."""
         prep_vals = self.parameters[0]
-        if any(i not in [0, 1] for i in prep_vals):
-            raise ValueError("BasisState parameter must consist of 0 or 1 integers.")
-
-        if (num_wires := len(self.wires)) != len(prep_vals):
-            raise ValueError("BasisState parameter and wires must be of equal length.")
```
Comment (soranjh) [inline pennylane/ops/qubit/state_preparation.py:111]: Why are these removed?

File: pennylane/templates/embeddings/basis.py [hunk 11]
```diff
@@ -15,17 +15,15 @@
...
-import numpy as np
 
-import pennylane as qml
-from pennylane.operation import AnyWires, Operation
-from pennylane.wires import Wires
+from pennylane.ops.qubit.state_preparation import BasisState
 
 
-class BasisEmbedding(Operation):
+# pylint: disable=missing-class-docstring
```
Comment (soranjh) [inline pennylane/templates/embeddings/basis.py:22]: Do we need this?

File: tests/templates/test_embeddings/test_basis.py [hunk 12]
```diff
@@ -245,19 +241,20 @@ def test_jax_jit(self, tol):
 
         features = jnp.array([0, 1, 0])
 
-        dev = qml.device("default.qubit", wires=3)
+        for dev_name in ["default.qubit", "lightning.qubit"]:
```
[the remark this answers is not in the document]
Comment (soranjh) [reply tests/templates/test_embeddings/test_basis.py:244]: Can we use `@pytest.mark.parametrize` here instead of the for loop?

**Review by soranjh — CHANGES_REQUESTED** [review]
Thanks @KetpuntoG, looks good overall. Left some comments and questions.

File: pennylane/ops/qubit/state_preparation.py [hunk 13]
```diff
@@ -20,24 +20,24 @@
...
 class BasisState(StatePrepBase):
-    r"""BasisState(n, wires)
+    r"""BasisState(features, wires)
     Prepares a single computational basis state.
 
     **Details:**
 
     * Number of wires: Any (the operation can act on any number of wires)
     * Number of parameters: 1
-    * Gradient recipe: None (integer parameters not supported)
```
Comment (soranjh) [inline pennylane/ops/qubit/state_preparation.py:40]: Why this is removed? Should it be updated instead?

File: pennylane/ops/qubit/state_preparation.py [hunk 14]
```diff
@@ -51,9 +51,8 @@ class BasisState(StatePrepBase):
         as :math:`U|0\rangle = |\psi\rangle`
 
     Args:
-        n (array): prepares the basis state :math:`\ket{n}`, where ``n`` is an
-            array of integers from the set :math:`\{0, 1\}`, i.e.,
-            if ``n = np.array([0, 1, 0])``, prepares the state :math:`|010\rangle`.
+        features (tensor_like): binary input of shape ``(len(wires), )``. For example, for ``features=np.array([0, 1, 0])`` or ``features=2`` (binary 010), the quantum system will be prepared in state :math:`|010 \rangle`.
```
Comment (soranjh) [inline pennylane/ops/qubit/state_preparation.py:54]:
```suggestion
        features (tensor_like): binary input of shape ``(len(wires), )``, e.g., for ``features=np.array([0, 1, 0])`` or ``features=2`` (binary 010), the quantum system will be prepared in state :math:`|010 \rangle`.
```

File: pennylane/ops/qubit/state_preparation.py [hunk 15]
```diff
@@ -69,15 +68,51 @@ class BasisState(StatePrepBase):
...
-    """int: Number of dimensions per trainable parameter of the operator."""
+        if isinstance(features, list):
+            features = qml.math.stack(features)
+
+        tracing = qml.math.is_abstract(features)
+
+        if qml.math.shape(features) == ():
+            if not tracing and features >= 2 ** len(wires):
+                raise ValueError(
+                    f"Features must be of length {len(wires)}, got features={features} which is >= {2 ** len(wires)}"
```
Comment (obliviateandsurrender) [inline pennylane/ops/qubit/state_preparation.py:81]:
```suggestion
                    f"Integer features must be < {{2 ** len(wires)} to have a feasible binary representation, got {features}"
```
This or some rewording along these lines.

File: pennylane/ops/qubit/state_preparation.py [hunk 16]
```diff
@@ -69,15 +68,51 @@ class BasisState(StatePrepBase):
...
+            raise ValueError(f"Features must be one-dimensional; got shape {shape}.")
+
+        n_features = shape[0]
+        if n_features != len(wires):
+            raise ValueError(
+                f"Features must be of length {len(wires)}; got length {n_features} (features={features})."
+            )
+
+        if not tracing:
+            features_list = list(qml.math.toarray(features))
```
Comment (obliviateandsurrender) [inline pennylane/ops/qubit/state_preparation.py:99]: Maybe we could use something like `qml.math.allclose(qml.math.unique(features), [0, 1])`

File: pennylane/ops/qubit/state_preparation.py [hunk 17]
```diff
@@ -96,33 +131,49 @@ def compute_decomposition(n: TensorLike, wires: WiresLike) -> list[Operator]:
...
+        [X(0)]
 
         """
-        return [BasisStatePreparation(n, wires)]
+
+        if not qml.math.is_abstract(features):
+            op_list = []
+            for wire, state in zip(wires, features):
+                if state == 1:
+                    op_list.append(qml.X(wire))
```
Comment (obliviateandsurrender) [inline pennylane/ops/qubit/state_preparation.py:142]:
```suggestion
            op_list = [qml.X(wire) for wire, state in zip(wires, features) if state == 1]
```

File: pennylane/ops/qubit/state_preparation.py [hunk 18]
```diff
@@ -69,15 +68,51 @@ class BasisState(StatePrepBase):
...
+        features = tuple(features) if isinstance(features, list) else features
+        return (features,), (self.wires,)
+
+    @classmethod
+    def _unflatten(cls, data, metadata) -> "BasisState":
+        return cls(data[0], wires=metadata[0])
 
     @staticmethod
-    def compute_decomposition(n: TensorLike, wires: WiresLike) -> list[Operator]:
+    def compute_decomposition(features: TensorLike, wires: WiresLike) -> list[Operator]:
```
Comment (obliviateandsurrender) [inline pennylane/ops/qubit/state_preparation.py:115]: We do not perform any sanity checks on `features` here?

File: pennylane/ops/qubit/state_preparation.py [hunk 19]
```diff
@@ -96,33 +131,49 @@ def compute_decomposition(n: TensorLike, wires: WiresLike) -> list[Operator]:
...
                 indices[wire_order.index(base_wire_label)] = value
 
-        ket = np.zeros((2,) * num_wires)
-        ket[tuple(indices)] = 1
+        if qml.math.get_interface(prep_vals_int) == "jax":
+            ket = math.array(math.zeros((2,) * num_wires), like="jax")
+            ket = ket.at[tuple(indices)].set(1)
+
+        else:
+            ket = math.zeros((2,) * num_wires)
```
Comment (obliviateandsurrender) [inline pennylane/ops/qubit/state_preparation.py:174]: We don't use `interface` for matching other non-numpy backends as well?

File: tests/ops/qubit/test_state_prep.py [hunk 20]
```diff
@@ -66,8 +66,8 @@ def test_BasisState_decomposition(self):
         ops2 = qml.BasisState(n, wires=wires).decomposition()
 
         assert len(ops1) == len(ops2) == 1
-        assert isinstance(ops1[0], qml.BasisStatePreparation)
-        assert isinstance(ops2[0], qml.BasisStatePreparation)
+        assert isinstance(ops1[0], qml.PauliX)
+        assert isinstance(ops2[0], qml.PauliX)
```
Comment (obliviateandsurrender) [inline tests/ops/qubit/test_state_prep.py:70]:
```suggestion
        assert isinstance(ops1[0], qml.X)
        assert isinstance(ops2[0], qml.X)
```

**Review by obliviateandsurrender — CHANGES_REQUESTED** [review]

File: pennylane/ops/qubit/state_preparation.py (hunk 10 above)
Comment (KetpuntoG) [reply pennylane/ops/qubit/state_preparation.py:111]: because BasisState now supports integers. Therefore, the input can be different from 0 and 1. Also, the dimensions do not necessarily have to match the wires

File: pennylane/ops/qubit/state_preparation.py (hunk 13 above)
Comment (KetpuntoG) [reply pennylane/ops/qubit/state_preparation.py:40]: removed only the brackets 👍

File: doc/releases/changelog-dev.md [hunk 21]
```diff
@@ -354,7 +354,16 @@
...
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply doc/releases/changelog-dev.md:366]: Combining them will make the changelog more contained, so I think we should do it.

File: pennylane/ops/qubit/state_preparation.py (hunk 18 above)
Comment (obliviateandsurrender) [reply pennylane/ops/qubit/state_preparation.py]: Technically we should, as we cannot be sure that `features` will always be correct and would give incorrect decomposition.

Comment (soranjh) [reply pennylane/ops/qubit/state_preparation.py]: We can stick to what the original templates were doing in this PR and make upgrades in the followup PR.

File: doc/releases/changelog-dev.md [hunk 22]
```diff
@@ -375,7 +375,10 @@
   [(#5978)](https://github.com/PennyLaneAI/pennylane/pull/5978)
 
 * `qml.AmplitudeEmbedding` has better support for features using low precision integer data types.
-[(#5969)](https://github.com/PennyLaneAI/pennylane/pull/5969)
+  [(#5969)](https://github.com/PennyLaneAI/pennylane/pull/5969)
+
+* `qml.BasisState` and `qml.BasisEmbedding` now works with jax.jit, lightning.qubit and now gives the correct decomposition.
```
Comment (soranjh) [inline doc/releases/changelog-dev.md:380]:
```suggestion
* `qml.BasisState` and `qml.BasisEmbedding` now works with jax.jit, lightning.qubit and give the correct decomposition.
```

**Review by soranjh — APPROVED** [review]
Thanks @KetpuntoG , looks good to go.

File: pennylane/ops/qubit/state_preparation.py [hunk 23]
```diff
@@ -96,33 +132,45 @@ def compute_decomposition(n: TensorLike, wires: WiresLike) -> list[Operator]:
...
 
     def state_vector(self, wire_order: Optional[WiresLike] = None) -> TensorLike:
         """Returns a statevector of shape ``(2,) * num_wires``."""
         prep_vals = self.parameters[0]
-        if any(i not in [0, 1] for i in prep_vals):
-            raise ValueError("BasisState parameter must consist of 0 or 1 integers.")
-
-        if (num_wires := len(self.wires)) != len(prep_vals):
-            raise ValueError("BasisState parameter and wires must be of equal length.")
+        prep_vals_int = math.cast(prep_vals, int)
```
Comment (obliviateandsurrender) [inline pennylane/ops/qubit/state_preparation.py:153]:
```suggestion
        prep_vals_int = math.cast(self.parameters[0], int)
```
We can directly do this.

**Review by obliviateandsurrender — APPROVED** [review]
A couple of final minor comments, but looks good otherwise.
